examples/step-56/step-56.cc

```diff
@@ -0,0 +1,1186 @@
+        return_value[1] =  - PI * sin (PI * x) * sin(PI * y) * sin (PI * z);
+        return_value[2] = PI * sin (PI * x) * cos (PI * y) * cos (PI * z);
+      }
+
+    return return_value;
+  }
+
+  // Implementation of $f$. See the introduction for more information.
+  template <int dim>
+  class RightHandSide : public Function<dim>
+  {
+  public:
+    RightHandSide () : Function<dim>(dim+1) {}
+
+    virtual double value (const Point<dim> &p,
+                          const unsigned int component = 0) const;
+  };
+
+  template <>
+  double
+  RightHandSide<2>::value (const Point<2> &p,
+                           const unsigned int component) const
+  {
+    Assert (component <= 2, ExcIndexRange(component,0,2+1));
+
+    using numbers::PI;
+    double x = p(0);
+    double y = p(1);
+    if (component == 0)
+      return PI * PI * sin(PI * x) + PI * cos(PI * x) * cos(PI * y);
+    if (component == 1)
+      return - PI * PI * PI * y * cos(PI * x) - PI * sin(PI * y) * sin(PI * x);
+    if (component == 2)
+      return 0;
+
+    return 0;
+  }
+
+  template <>
+  double
+  RightHandSide<3>::value (const Point<3>   &p,
+                           const unsigned int component) const
+  {
+    Assert (component <= 3, ExcIndexRange(component,0,3+1));
+
+    using numbers::PI;
+    double x = p(0);
+    double y = p(1);
+    double z = p(2);
+    if (component == 0)
+      return 2 * PI * PI * sin(PI * x) + PI * cos(PI * x) * cos(PI * y) * sin(PI * z);
+    if (component == 1)
+      return  - PI * PI * PI * y * cos (PI * x) + PI * (-1) * sin(PI * y)*sin(PI * x)*sin(PI * z);
+    if (component == 2)
+      return - PI * PI * PI * z * cos (PI * x) + PI * cos(PI * z)*sin(PI * x)*cos(PI * y);
+    if (component == 3)
+      return 0;
+
+    return 0;
+  }
+
+  // Sadly, we need a separate function for the boundary conditions
+  // to be used in the geometric multigrid. This is because it needs
+  // to be a function with $dim$ components, whereas Solution has
+  // $dim+1$ components. Rather than copying the implementation of
+  // Solution, we forward the calls to Solution::value. For that we need
+  // an instance of the class Solution, which you can find as a private
+  // member.
+  template <int dim>
+  class BoundaryValuesForVelocity : public Function<dim>
+  {
+  public:
+    BoundaryValuesForVelocity () : Function<dim>(dim) {}
+
+    virtual double value (const Point<dim>   &p,
+                          const unsigned int  component = 0) const;
+
+  private:
+    Solution<dim> solution;
+  };
+
+
+  template <int dim>
+  double
+  BoundaryValuesForVelocity<dim>::value (const Point<dim>  &p,
+                                         const unsigned int component) const
+  {
+    Assert (component < this->n_components,
+            ExcIndexRange (component, 0, this->n_components));
+
+    return solution.value(p, component);
+  }
+
+
+
+  // @sect3{ASPECT BlockSchurPreconditioner}
```

Comment:
since the previous comments of mine and JP are hidden now, i would like to bring again the discussion about re-implementing the Schur preconditioner. I think it makes a lot of sense to show the user how to utilize classes/functions provided by `deal.ii`, in this context [SchurComplement](https://www.dealii.org/8.4.0/doxygen/deal.II/group__LAOperators.html#ga1cb1a9bb9e30355214718b66ac71d107). That will (i) save some 100 lines of code in this tutorial; (ii) this tutorial can serve as another unit test for this functionality.